File: app/src/main/java/org/thoughtcrime/securesms/service/AccountAuthenticatorService.java
package org.thoughtcrime.securesms.service;

import android.accounts.AbstractAccountAuthenticator;
import android.accounts.Account;
import android.accounts.AccountAuthenticatorResponse;
import android.accounts.NetworkErrorException;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.os.IBinder;

public class AccountAuthenticatorService extends Service {

  private static AccountAuthenticatorImpl accountAuthenticator = null;

  @Override
  public IBinder onBind(Intent intent) {
    if (intent.getAction().equals(android.accounts.AccountManager.ACTION_AUTHENTICATOR_INTENT)) {
      return getAuthenticator().getIBinder();
    } else {
      return null;
    }
  }

  private synchronized AccountAuthenticatorImpl getAuthenticator() {
    if (accountAuthenticator == null) {
      accountAuthenticator = new AccountAuthenticatorImpl(this);
    }

    return accountAuthenticator;
  }

  private static class AccountAuthenticatorImpl extends AbstractAccountAuthenticator {

    public AccountAuthenticatorImpl(Context context) {
      super(context);
    }

    @Override
    public Bundle addAccount(AccountAuthenticatorResponse response, String accountType, String authTokenType,
                             String[] requiredFeatures, Bundle options)
        throws NetworkErrorException
    {
      return null;
    }

    public Bundle confirmCredentials(AccountAuthenticatorResponse response, Account account, Bundle options) {
      return null;
    }

    @Override
    public Bundle editProperties(AccountAuthenticatorResponse response, String accountType) {
      return null;
    }

    @Override
    public Bundle getAuthToken(AccountAuthenticatorResponse response, Account account, String authTokenType,
                               Bundle options) throws NetworkErrorException {
      return null;
    }

    @Override
    public String getAuthTokenLabel(String authTokenType) {
      return null;
    }

    @Override
    public Bundle hasFeatures(AccountAuthenticatorResponse response, Account account, String[] features)
        throws NetworkErrorException {
      return null;
    }

    @Override
    public Bundle updateCredentials(AccountAuthenticatorResponse response, Account account, String authTokenType,
                                    Bundle options) {
      return null;
    }
  }
}

File: contacts/app/src/main/java/org/signal/contactstest/AccountAuthenticatorService.kt
package org.signal.contactstest

import android.accounts.AbstractAccountAuthenticator
import android.accounts.Account
import android.accounts.AccountAuthenticatorResponse
import android.accounts.AccountManager
import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.os.IBinder

class AccountAuthenticatorService : Service() {
  companion object {
    private var accountAuthenticator: AccountAuthenticatorImpl? = null
  }

  override fun onBind(intent: Intent): IBinder? {
    return if (intent.action == AccountManager.ACTION_AUTHENTICATOR_INTENT) {
      getOrCreateAuthenticator().iBinder
    } else {
      null
    }
  }

  @Synchronized
  private fun getOrCreateAuthenticator(): AccountAuthenticatorImpl {
    if (accountAuthenticator == null) {
      accountAuthenticator = AccountAuthenticatorImpl(this)
    }
    return accountAuthenticator as AccountAuthenticatorImpl
  }

  private class AccountAuthenticatorImpl(context: Context) : AbstractAccountAuthenticator(context) {
    override fun addAccount(response: AccountAuthenticatorResponse, accountType: String, authTokenType: String, requiredFeatures: Array<String>, options: Bundle): Bundle? {
      return null
    }

    override fun confirmCredentials(response: AccountAuthenticatorResponse, account: Account, options: Bundle): Bundle? {
      return null
    }

    override fun editProperties(response: AccountAuthenticatorResponse, accountType: String): Bundle? {
      return null
    }

    override fun getAuthToken(response: AccountAuthenticatorResponse, account: Account, authTokenType: String, options: Bundle): Bundle? {
      return null
    }

    override fun getAuthTokenLabel(authTokenType: String): String? {
      return null
    }

    override fun hasFeatures(response: AccountAuthenticatorResponse, account: Account, features: Array<String>): Bundle? {
      return null
    }

    override fun updateCredentials(response: AccountAuthenticatorResponse, account: Account, authTokenType: String, options: Bundle): Bundle? {
      return null
    }
  }
}
